### PR TITLE
feature: support HTTP/3 testing.

### PR DESCRIPTION
**Pull request by zhuizhuhaomeng**

### CODE REVIEW

File: lib/Test/Nginx/Socket.pm [hunk 1]
```diff
@@ -1783,16 +1783,33 @@ sub send_http2_req ($$$) {
         warn "running cmd @$cmd";
     }
 
+
```
Comment (xiaocang) [inline lib/Test/Nginx/Socket.pm:1786]: maybe we should remove the extra blank line here

File: lib/Test/Nginx/Socket.pm [hunk 2]
```diff
@@ -1783,16 +1783,33 @@ sub send_http2_req ($$$) {
...
     my $ok = IPC::Run::run($cmd, \(my $in), \(my $out), \(my $err),
                            IPC::Run::timeout($timeout));
 
+    my @cmd_copy = @$cmd;
```
Comment (xiaocang) [inline lib/Test/Nginx/Socket.pm:1790]: comment out the debug code here

File: lib/Test/Nginx/Socket.pm [hunk 3]
```diff
@@ -2254,24 +2280,23 @@ sub gen_curl_cmd_from_req ($$) {
...
+
+        push @args, '-H', $h;
     }
 
     if ($req =~ m{\G(.+)}gcsm) {
         #warn "!! POST body data len: ", length($1);
         if (!$found_content_type) {
             push @args, "-H", 'Content-Type: ';
         }
-        push @args, '--data-binary', $1;
```
Comment (xiaocang) [inline lib/Test/Nginx/Socket.pm:2274]: We should assign the match result to a local variable, if it is not used immediately.

```
my $body = $1;
```

File: lib/Test/Nginx/Socket.pm [hunk 4]
```diff
@@ -2254,24 +2280,23 @@ sub gen_curl_cmd_from_req ($$) {
...
+        my $filename = html_dir() . "/curl.data.bin";
+        push @args, '--data-binary', '@' . $filename;
+        open my $fh, ">", $filename or die("Could not open file. $!");
```
Comment (xiaocang) [inline lib/Test/Nginx/Socket.pm:2297]: Calling the `die` function here does not require parentheses

File: lib/Test/Nginx/Socket.pm [hunk 5]
```diff
@@ -2280,6 +2305,7 @@ sub gen_curl_cmd_from_req ($$) {
     }
 
     push @args, '--connect-timeout', $timeout;
+    push @args, '--max-time', $timeout;
```
Comment (xiaocang) [inline lib/Test/Nginx/Socket.pm:2308]: some comments for this option should be added here

File: lib/Test/Nginx/Util.pm [hunk 6]
```diff
@@ -2689,6 +2743,12 @@ sub use_http2 ($) {
             return undef;
         }
 
+        if (defined $block->http3) {
+            warn "WARNING: ", $block->name, " - explicitly rquires HTTP/3, so will not use HTTP/2\n";
```
Comment (xiaocang) [inline lib/Test/Nginx/Util.pm:2747]: typo here: 'rquires' -> 'requires'

File: lib/Test/Nginx/Util.pm [hunk 7]
```diff
@@ -2703,4 +2763,78 @@ sub use_http2 ($) {
...
+            return undef;
+        }
+
+        if (!defined $block->request) {
+            $block->set_value("test_nginx_enabled_http3", 0);
+            return undef;
+        }
+
+        if (!ref $block->request && $block->request =~ m{HTTP/1\.0}s) {
+            warn "WARNING: ", $block->name, " - explicitly rquires HTTP 1.0, so will not use HTTP/3\n";
```
Comment (xiaocang) [inline lib/Test/Nginx/Util.pm:2816]: Ditto

File: lib/Test/Nginx/Util.pm [hunk 8]
```diff
@@ -2703,4 +2763,78 @@ sub use_http2 ($) {
...
+            $block->set_value("test_nginx_enabled_http3", 0);
+            return undef;
+        }
+
+        if (defined $block->http2) {
+            warn "WARNING: ", $block->name, " - explicitly rquires HTTP/2, so will not use HTTP/3\n";
```
Comment (xiaocang) [inline lib/Test/Nginx/Util.pm:2822]: Ditto

File: lib/Test/Nginx/Util.pm [hunk 9]
```diff
@@ -2703,4 +2763,78 @@ sub use_http2 ($) {
...
+        $block->set_value("test_nginx_enabled_http3", 1);
+
+        if (!$LoadedIPCRun) {
+            require IPC::Run;
+            $LoadedIPCRun = 1;
+        }
+        return 1;
+    }
+
+    $block->set_value("test_nginx_enabled_http3", 0);
```
Comment (xiaocang) [inline lib/Test/Nginx/Util.pm:2836]: set_value to test_nginx_enabled_http3 should be `1` here

**Review by xiaocang — CHANGES_REQUESTED** [review]

File: lib/Test/Nginx/Util.pm (hunk 9 above)
Comment (zhuizhuhaomeng) [reply lib/Test/Nginx/Util.pm:2837]: 0 is right.
use_http3 is used to test if should use http3 for the test case, not to enable http3.

**Review by xiaocang — APPROVED** [review]
LGTM

**Pull request merged**